src/session/content/chat_history.rs

```diff
@@ -163,7 +165,29 @@ impl ChatHistory {
                 _ => self.action_set_enabled("chat-history.view-info", false),
             }
 
-            let selection = gtk::NoSelection::new(Some(&chat.history()));
+            let list = gio::ListStore::new(gio::ListModel::static_type());
+            list.append(&chat.history());
+
+            if matches!(chat.type_(), ChatType::Supergroup(data) if data.is_channel) {
+                if self_.sponsored_message_list.borrow().is_none() {
+                    let sponsored_message_list = SponsoredMessageList::new();
+                    self_
+                        .sponsored_message_list
+                        .replace(Some(sponsored_message_list));
+                }
+
+                if let Some(sponsored_message_list) = self_.sponsored_message_list.borrow().as_ref()
+                {
+                    sponsored_message_list.clear();
+                    sponsored_message_list.fetch(chat);
+                    list.append(sponsored_message_list);
+                } else {
+                    unreachable!("The sponsored message list is expected to exist here");
+                }
```

Comment:
It seems to me that `if let` statement isn't necessary because the sponsored message list is always `some` at this point.
I think we want something like this (just one borrow):

```suggestion
                let mut sponsored_message_list_ref = self_.sponsored_message_list.borrow_mut();
                let sponsored_message_list = match *sponsored_message_list_ref {
                    None => {
                        let sponsored_message_list = SponsoredMessageList::new();
                        *sponsored_message_list_ref = Some(sponsored_message_list);
                        sponsored_message_list_ref.as_ref().unwrap()
                    }
                    Some(ref sponsored_message_list) => {
                        sponsored_message_list.clear();
                        sponsored_message_list
                    }
                };

                sponsored_message_list.fetch(chat);
                list.append(sponsored_message_list);
```

Reply:
Yeah, it looks much better, thanks!